### PR TITLE
a way for async exceptions to get treated as such

### PR DESCRIPTION
**Pull request by flyingmachine**
This addresses #11108 

When an exception occurs on an API endpoint that returns async responses, the frontend doesn't "know" about it. This is because async responses always return a 200 status, and the frontend thinks that a 200 status means the request was successful.

This PR honestly feels like a hack. I think it would be better to establish a single error representation and always return. Going by [api-exception-response](https://github.com/metabase/metabase/blob/master/src/metabase/middleware/exceptions.clj#L43) it seems like, at the very least, response bodies can be either strings or maps.

### CODE REVIEW

**Comment by flyingmachine** [comment]
@paulrosenzweig @tlrobinson the FE tests are failing on this but before fixing them I'd love to hear whether you think this approach is a good one :)

**Comment by paulrosenzweig** [comment]
I think it'd be best to submit to the will of HTTP and using status codes, but from our chat yesterday it sounds like that was a difficult thing to restructure. Given that, I think this is the place to hook in.

Are there only a few routes where we can expect this type of error? If so, rather than have this setup for every API call, I'd prefer to explicitly tell those API methods to expect this kind of error. For example, to have the card creation endpoint check for an error we could pass this option:

`POST("/api/card", {bodyHasError: (body) => body.async_status === "error"})`

**Comment by flyingmachine** [comment]
Yeah - my understanding is that the way the async response works is that the serve immediately responds with a 200 and then streams a body. The streaming is not a function of the body's size, it's a consequence of handling the request asynchronously in the backend.

I did a quick and dirty survey of how many routes we can expect this kind of error from, and it looks like it's `/api/card` and `/api/dataset` endpoints

**Comment by flyingmachine** [comment]
@paulrosenzweig would it be possible to hand this off to you for any further frontend work?

**Comment by paulrosenzweig** [comment]
I added an option called `bodyHasError` and set it up for Card create/update. It looks like those are the only routes that use the async handler, but if I misread, we can add it to the other routes.

If I add `(throw (Exception. "oh no"))` to `save-new-card-async!`, I see the error message after trying to save:
image

**Comment by camsaul** [comment]
I don't think it makes a ton of sense to have async error responses be either strings or maps, especially since I'm 95% sure it comes back with `Content-Type: application/json` headers either way. We should return them in some sort of standard error map instead.

What we really need is to come up with a few standard REST API patterns we can agree on and write them up in a wiki page

File: frontend/src/metabase/lib/api.js [hunk 1]
```diff
@@ -179,20 +179,24 @@ export class Api extends EventEmitter {
           try {
             body = JSON.parse(body);
           } catch (e) {}
-          if (xhr.status >= 200 && xhr.status <= 299) {
+          let status_code = xhr.status;
```
Comment (tlrobinson) [inline frontend/src/metabase/lib/api.js]: We don't typically use snake case in JavaScript. I'd just call it `status`.

File: frontend/src/metabase/lib/api.js [hunk 2]
```diff
@@ -179,20 +179,24 @@ export class Api extends EventEmitter {
...
+          if (status_code === 202) {
```
Comment (tlrobinson) [inline frontend/src/metabase/lib/api.js]: Might want to also check that `body.status_code` is not null/undefined.

**Review by tlrobinson — CHANGES_REQUESTED** [review]
Maybe we should use `_status` (or `_status_code`) as the special property? Rails used `_method` for a similar purpose, for example.

**Comment by flyingmachine** [comment]
it would be good to implement deeper structural changes to always return a 202 status for async requests, but I'd prefer to leave that for a separate PR.

**Comment by flyingmachine** [comment]
@tlrobinson this is ready for another look :)

**Review by camsaul — APPROVED** [review]
Cool LGTM.

Could you write some of this down in https://github.com/metabase/metabase/wiki/Metabase-REST-API-Patterns ? It will save us some time in the long run, when someone new joins we'll have to explain the pattern anyway so better to do it in one canonical place

**Comment by camsaul** [comment]
Also (not sure if there's already an open issue) can you open an issue for returning the 202 response for all async API requests, so we don't forget to do it? 🙏

**Comment by flyingmachine** [comment]
it looks like I need @tlrobinson to approve, or for an admin to merge :'(

File: frontend/src/metabase/lib/api.js [hunk 3]
```diff
@@ -179,20 +179,28 @@ export class Api extends EventEmitter {
           try {
             body = JSON.parse(body);
           } catch (e) {}
-          if (xhr.status >= 200 && xhr.status <= 299) {
+          let status = xhr.status;
+          if (
+            status === 202 &&
+            body._status !== null &&
```
Comment (tlrobinson) [inline frontend/src/metabase/lib/api.js]: This is super minor but the idiomatic way to check for `null` or `undefined` in JS is `!= null` (the only reason you should ever use `==` or `!=` over `===` and `!==`)

**Review by tlrobinson — APPROVED** [review]
Looks good, I made one small tweak, we can merge once it passes CI.

**Pull request merged**